src/data/copying_prob_src.rs

```diff
@@ -0,0 +1,154 @@
+    {
+        let input_dims = Dim4::new(&[batch_size, input_size, bptt_unroll, 1]);
+        //let output_dims = Dim4::new(&[batch_size, output_size, bptt_unroll, 1]);
+        let train_samples = 0.7 * max_samples as f32;
+        let test_samples = 0.2 * max_samples as f32;
+        let validation_samples = 0.1 * max_samples as f32;
+        CopyingProblemSource {
+            params : DataParams {
+                input_dims: input_dims,
+                target_dims: input_dims,
+                dtype: dtype,
+                normalize: false,
+                shuffle: false,
+                current_epoch: Cell::new(0),
+                num_samples: max_samples,
+                num_train: train_samples as u64,
+                num_test: test_samples as u64,
+                num_validation: Some(validation_samples as u64),
+            },
+            iter: Cell::new(0),
+            offset: Cell::new(0.0f32),
+            bptt_unroll: bptt_unroll,
+            seq_size: seq_size,
+        }
+    }
+
+    fn generate_input(&self, batch_size: u64, input_size: u64, bptt_unroll: u64, seq_size: u64) -> Array {
+
+        let between = Range::new(0,input_size-1);
+        let mut rng = rand::thread_rng();
+        
+        let mut vec_total = Vec::new();
+        let vec = vec!(0f32; input_size as usize);
+
+        for i in 0..seq_size {
+            for j in 0..batch_size {
+                let index = between.ind_sample(&mut rng) as usize;
+                let mut vec_temp = vec.clone();
+                vec_temp[index] = 1f32;
+                vec_total.append(&mut vec_temp);
+            }
+        }
+
+        for i in 0..bptt_unroll-2*seq_size-1 {
+            for j in 0..batch_size {
+                let mut vec_temp = vec.clone();
+                vec_temp[(input_size-2) as usize] = 1f32;
+                vec_total.append(&mut vec_temp);
+            }
+        }
+
+        for j in 0..batch_size {
+            let mut vec_temp = vec.clone();
+            vec_temp[(input_size-1) as usize] = 1f32;
+            vec_total.append(&mut vec_temp);
+        }
+
+        for i in 0..seq_size {
+            for j in 0..batch_size {
+                let mut vec_temp = vec.clone();
+                vec_temp[(input_size-2) as usize] = 1f32;
+                vec_total.append(&mut vec_temp);
+            }
+        }
+
+        let ar_dims = Dim4::new(&[input_size, batch_size, bptt_unroll, 1]);
+        af::transpose(&utils::vec_to_array::<f32>(vec_total, ar_dims), false)
+    }
+
+    fn generate_target(&self, input: &Array, batch_size: u64, input_size: u64, bptt_unroll: u64, seq_size: u64) -> Array {
+        let mut first = af::constant(0f32, af::Dim4::new(&[batch_size, input_size, bptt_unroll-seq_size,1]));
+        let ones = af::constant(1f32, af::Dim4::new(&[batch_size, 1, 1, 1]));
+
+        for i in 0..bptt_unroll-seq_size {
+            first = af::set_slice(&first, &af::set_col(&af::slice(&first, i), &ones, input_size-2), i);
```

Comment:
slices are VERY VERY slow as there is no contiguous allocation in GPU memory. This should be done in host memory (i.e. in a Vector) and then only after it is done converted to an array